scheduling-utils/src/handshake/server.rs

```diff
@@ -0,0 +1,323 @@
+    pub fn accept(&mut self) -> Result<AgaveSession, AgaveHandshakeError> {
+        // Clear the buffer.
+        self.buffer_len = 0;
+
+        // Wait for next stream.
+        let (mut stream, _) = self.listener.accept()?;
+        stream.set_read_timeout(Some(HANDSHAKE_TIMEOUT))?;
+
+        match self.handle_logon(&mut stream) {
+            Ok(session) => Ok(session),
+            Err(err) => {
+                let reason = err.to_string();
+                let reason_len = u8::try_from(reason.len()).unwrap_or(u8::MAX);
+
+                let buffer_len = 2usize.checked_add(usize::from(reason_len)).unwrap();
+                self.buffer[0] = LOGON_FAILURE;
+                self.buffer[1] = reason_len;
+                self.buffer[2..buffer_len]
+                    .copy_from_slice(&reason.as_bytes()[..usize::from(reason_len)]);
+
+                stream.set_nonblocking(true)?;
+                // NB: Caller will still error out even if our write fails so it's fine to ignore the
+                // result.
+                let _ = stream.write(&self.buffer[..buffer_len])?;
+
+                Err(err)
+            }
+        }
+    }
+
+    fn handle_logon(
+        &mut self,
+        stream: &mut UnixStream,
+    ) -> Result<AgaveSession, AgaveHandshakeError> {
+        // Read the logon message.
+        let handshake_start = Instant::now();
+        while self.buffer_len < self.buffer.len() {
+            let read = stream.read(&mut self.buffer[self.buffer_len..])?;
+            if read == 0 {
+                return Err(AgaveHandshakeError::EofDuringHandshake);
+            }
+
+            // SAFETY: We cannot read a value greater than buffer.len() which itself is a usize.
+            self.buffer_len = self.buffer_len.checked_add(read).unwrap();
+
+            if handshake_start.elapsed() > HANDSHAKE_TIMEOUT {
+                return Err(AgaveHandshakeError::Timeout);
+            }
+        }
+
+        // Ensure exact version match, version will be bumped any time a backwards incompatible
+        // change is made to handshake/shared memory objects.
+        let version = u64::from_le_bytes(self.buffer[..8].try_into().unwrap());
+        if version != VERSION {
+            return Err(AgaveHandshakeError::Version {
+                server: VERSION,
+                client: version,
+            });
+        }
+
+        // Read the logon message, cannot panic as we ensure the correct buf size at compile time
+        // (hence the const just below).
+        const LOGON_END: usize = 8 + core::mem::size_of::<ClientLogon>();
+        let logon = ClientLogon::try_from_bytes(&self.buffer[8..LOGON_END]).unwrap();
+
+        // Put a hard limit of 64 worker threads for now.
+        if !(1..=MAX_WORKERS).contains(&logon.worker_count) {
+            return Err(AgaveHandshakeError::WorkerCount(logon.worker_count));
+        }
+
+        // Hard limit allocator handles to 128.
+        if logon.allocator_handles > MAX_ALLOCATOR_HANDLES {
```

Comment:
let's also verify it's at least one?
messages in the tpu_to_pack queue make no sense without the allocator, and agave (if receiving txs) will fill up the allocator and fall back eventually if the mem doesn't get freed by external.